pkg/testutils/mock_queue.go

```diff
@@ -79,9 +79,14 @@ func (q *MockWorkQueue) ExpectAdds(diff int) {
 // Wait waits until the expected amount of ExpectedAdds has happened.
 // It will not block if there were no expectations set.
 func (q *MockWorkQueue) Wait() {
-	if q.addWG != nil {
-		q.addWG.Wait()
+	q.wgLock.Lock()
+	wg := q.addWG
+	q.wgLock.Unlock()
```

Comment:
Pointers assignment and read are usually atomic by default. Do I miss something?

Reply:
Not sure what you mean. The concurrency model of golang does not guarantee you synchronicity betwenn go routines. It does not matter how something is written to memory. If you don't modify or access something between locks (or use channels or waitgroups which are meant for concurrent access) you have to use locks to see the same thing on different go routines.

Reply:
To explain this a little bit more: THe waitgroup itself is go-routine save, but first I have to get it from a location which I need to access in a locked fashion. If I don't do that, one go routine could see `nil`, the next one  the correct WG and a third one an old WG.

Reply:
If I understand correctly now, the usages of locks in this PR are placed in order to implement a memory barrier.
This is usually needed only if one tries to access shared data directly instead of through channels.

Anyway, I think there are dedicated sync actions that can be used to avoid locking, like [Value](https://golang.org/pkg/sync/atomic/#Value) from "sync/atomic". Pointers and integers can use dedicated ones, like [LoadPointer](https://golang.org/pkg/sync/atomic/#LoadPointer).

But generally speaking, I find this very confusing and future us may not understand where it is coming from. If we cannot change it to use some better way that will avoid data sharing, we better explain this special sync usage in the commit message.

Reply:
yes we want to access it directly. I don't think that channels make here anything simpler. I don't think that your suggestions make in this case anything simpler.

Reply:
I just think `wg := atomic.LoadPointer(&q.addWG)` and `atomic.StorePointer(&q.addWG, nil)` is an oneliner and easier to read (especial when the mutex is used as a memory barrier and not its mutual exclusion feature).

Even a `_ := atomic.LoadPointer(q.addWG)` is enough probably.

Reply:
I really see no reason here to work with unsafe pointers and/or type casting.

Reply:
And I see no reason to use a lock/unlock on an operation that has an equivalent atomic operation.
The need for this is already a smell that something is really bad here, limiting the "noise" is preferred IMO.

If the existing usage is kept, I would appreciate an explanation in the commit so we will not have to ask the same in the future.

Reply:
There is no equivalent atomic operation. lock/unlock is the standard way (if channels are not a good option) if you are not writing low level algorithms. There is no reason to write that to a commit message. You can not just pass in an arbitrary pointer to an `*unsafe.Pointer` argument.

Reply:
If I understand correctly, the "unsafe" part here is that the pointer type is not validated through the type checking.
But we are here in a test stub/mock, doing something unorthodox anyway, so I do not see the problem. The equivalent part is that reading or writing to a pointer can be done atomically without the need of a lock and you claim we still need to use a lock.

The commit needs to explain why the lock is needed and why it was preferred over the use of the atomic pointer write/read option. It may be obvious to some, but not that obvious to others (e.g. /me).
The use of a lock, from my point of view is crystal clear when there is a need to protect a critical section, but not when the section is one operation which can be implemented using atomic alternatives (e.g. pointer set/read, increment, assignment, etc).

Reply:
> If I understand correctly, the "unsafe" part here is that the pointer type is not validated through the type checking.
> But we are here in a test stub/mock, doing something unorthodox anyway, so I do not see the problem. The equivalent part is that reading or writing to a pointer can be done atomically without the need of a lock and you claim we still need to use a lock.

I mostly mean that your examples don't even compile.



> The commit needs to explain why the lock is needed and why it was preferred over the use of the atomic pointer write/read option. It may be obvious to some, but not that obvious to others (e.g. /me).
> The use of a lock, from my point of view is crystal clear when there is a need to protect a critical section, but not when the section is one operation which can be implemented using atomic alternatives (e.g. pointer set/read, increment, assignment, etc).

Code reviews are a great chance to learn by doing research outside of the review first. This way one can learn pretty quickly and does not have to get explained idiomatic go in code reviews.

Reply:
I really do not understand where you are going with this.
We have discussed until now the conceptual idea here, not the exact implementation at the compilation level.
If we are now moving to an exact implementation, then lets move focus there and avoid assumptions about what the reviewers have done or not.

To conclude this discussion, this is my suggestion (that compiles :)):
```golang
func (q *MockWorkQueue) Wait() {
	addWG_upaddr := (*unsafe.Pointer)(unsafe.Pointer(&q.addWG))
	wq := (*sync.WaitGroup)(atomic.LoadPointer(addWG_upaddr))
	if wq != nil {
		wq.Wait()
		atomic.StorePointer(addWG_upaddr, unsafe.Pointer(nil))
	}
}
```

The unsafe & atomic can be abstracted behind functions to reduce the noise from a reader.

Reply:
We have to be more efficient here in these reviews. The time spent on this looks ridiculous given we're talking about a trivial synchronization change for a unit test that functionally appears fine.

If @EdDev authored the patch and used this atomic pointer stuff, i'd just take a look and move on. I'd do the same with @rmohr's mutex.

At the absolute **most** if i saw the change functionally worked but i liked an alternative more, i'd offer the alternative as a suggestion, but that would be the end of it... it would be a suggestion, then we'd all move on because it doesn't matter enough to discuss further.

Reply:
> To conclude this discussion, this is my suggestion (that compiles :)):

Great, now you can probably see how difficult it is to simply using a lock on all places (you have to use it on the other places too). If you can't see how difficult this is compared to the other, I really can't help you.

Reply:
> We have to be more efficient here in these reviews. The time spent on this looks ridiculous given we're talking about a trivial synchronization change for a unit test that functionally appears fine.


Agreed.

> 
> If @EdDev authored the patch and used this atomic pointer stuff, i'd just take a look and move on. I'd do the same with @rmohr's mutex.

> 
> At the absolute **most** if i saw the change functionally worked but i liked an alternative more, i'd offer the alternative as a suggestion, but that would be the end of it... it would be a suggestion, then we'd all move on because it doesn't matter enough to discuss further.

Exactly